libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,265 @@
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def get_by_query(
+        self,
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        sort: Optional[Union[Sort, List[Sort]]] = None,
+        **kwargs: Any,
+    ) -> Iterable[T]:
+        """Get items by query."""
+
+    @abc.abstractmethod
+    def delete_by_query(
+        self,
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        sort: Optional[Union[Sort, List[Sort]]] = None,
+        **kwargs: Any,
+    ) -> Iterable[DeleteResponse]:
+        """Delete items by query."""
+        # Careful with halloween problem
+
+
+class Hit(Generic[T], TypedDict):
+    # Should subclass from content?
+    id: str
+    score: float
+    # Optional source
+    source: Optional[T]
+    metadata: Dict[str, Any]
+
+
+class QueryResponse(TypedDict):
+    """A retrieval result."""
+
+    # Free form metadata for vectorstore providers
+    metadata: Dict[str, Any]
+    hits: List[Hit]
+
+
+class UpsertDocument(TypedDict):
+    """A structured representation of a data item to upsert.
+
+    This allows providing additional information such as the vector.
+    """
+
+    id: str  # An override of the ID for the document
+    document: Document
+    vector: List[float]
+
+
+class VectorStore(BaseIndex[Document]):
+    @abc.abstractmethod
+    def query(
+        self,
+        query: Union[str, List[float]],
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        sort: Optional[Union[Sort, List[Sort]]] = None,
+        search_method: Optional[str] = None,
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Any,
+    ) -> QueryResponse[Document]:
+        """Query for items."""
+
+    @abc.abstractmethod
+    def upsert_with_vector(
+        self,
+        data_stream: Iterable[UpsertDocument],
+        /,
+        **kwargs: Any,
+    ) -> Iterable[UpsertResponse]:
+        """Upsert vectors by id."""
+
+
+# -----------------------------
+# Rely a bit on generics for parameterization
+
+Q = TypeVar("Q")
+
+
+class UpsertData(TypedDict, Generic[T]):
+    """A structured representation of a data item to upsert.
+
+    This allows providing additional information such as the vector.
+    """
+
+    content: T
+    vector: List[float]
+
+
+class VectorStoreGeneric(Generic[Q, T], BaseIndex[T]):
+    @abc.abstractmethod
+    def query(
+        self, query: Q, config: Optional[RunnableConfig] = None, **kwargs: Any
```

Comment:
This version parameterizes the query, so we could define something like


```python
class Query(TypedDict):
  query: Union[str, List[float]]
  filter: ...

```
And allow each vectorstore to vary the representation if they need to